File: 2699-count-the-number-of-fair-pairs/count-the-number-of-fair-pairs.go
func countFairPairs(nums []int, lower int, upper int) int64 {
    sort.Ints(nums);

    return count(nums, upper) - count(nums, lower - 1);
}

func count(nums []int, x int) int64 {
    left, right := 0, len(nums) - 1;
    var result int64 = 0;

    for left < right {
        if nums[left] + nums[right] > x {
            right--;
        } else {
            result += int64(right - left);
            left++;
        }
    }

    return result;
}

File: 2699-count-the-number-of-fair-pairs/count-the-number-of-fair-pairs.java
class Solution {
    public long countFairPairs(int[] nums, int lower, int upper) {
        Arrays.sort(nums);

        return count(nums, upper) - count(nums, lower - 1);
    }

    public long count(int[] nums, int x) {
        int left = 0, right = nums.length - 1;
        long result = 0;

        while (left < right) {
            if (nums[left] + nums[right] > x) right--;
            else result += right - left++;
        }

        return result;
    }
}